### ISSUE TITLE
Blaze as a statistical and ML modeling front end DSL

### ISSUE THREAD

**Issue by datnamer** [blaze/blaze#1450]
I had floated the idea to @kwmsmith of using blaze as a general front end for statistical modeling and machine learning DSLs, and he asked me for a write up. 

Unfortunately this is going to be a bit truncated and less polished  vs what i was planning, but I wanted to start the discussion because i was prompted  by a new Julia package, [Flow.jl](https://github.com/MikeInnes/Flow.jl), that pretty much describes what I had in mind. 

There are a variety of ad hoc python DSLs that allow expression of statistical and stochastic relationships that can be translated to lower level operations. 

Problem is, there is no shared infrastructure so every package has to reinvent the wheel and they require hacks (context managers, string programming etc) because they are  limited by python's syntax and eagerly evaluated semantics:

Examples of packages and constraints: [Patsy](http://patsy.readthedocs.org/en/latest/formulas.html) is not easily extensible for things like operations and multiple variables on the left hand side of the formula and Pymc 3 is restricted in extending it for  lme4 style mixed models. 

Both [Pymc 3](https://github.com/pymc-devs/pymc3/blob/master/pymc3/examples/disaster_model.ipynb) and now  [Keras](https://gist.github.com/fchollet/314085fffa200de9c3da) have their own such DSL graph front ends. 

I wonder if blaze can be used as a more flexible common infrastructure for these uscases (with more power to write symbolically and   dispatch across multiple datashapes and backends by offloading to dask, dynd etc

Or perhaps the fact that blaze is restricted to an expression tree vs DAG or cyclical graph might make it not a good fit for this usecase. 

cc @stevenmanton @twiecki, @njsmith and @richardotis who expressed interest or might be interested in this.

**Comment by richardotis** [comment]
I'd like to add some thoughts on where I see capability gaps in translating symbolic models to efficient numeric functions, particular when dealing with gradients and Hessians. This is based on my experience writing and maintaining [pycalphad](http://pycalphad.readthedocs.org/en/latest/).

The [code generation features in SymPy](http://matthewrocklin.com/blog/work/2013/03/19/SymPy-Theano-part-1) have been able to meet all my needs in terms of expressing conceptually simple models (basic arithmetic, logs, exps, but with 1000s of ops) and allowing run-time construction of the expressions. (I know their team is working on making it even better, see sympy/sympy#10486.)

The system is flexible enough to allow me to drop in custom code generators such as the now-upstream [`NumPyPrinter`](https://github.com/sympy/sympy/pull/9657), and the generated functions even allow for broadcasting, which is important for my use case, where I'm often mapping `R^n -> R`, where n is roughly 10, but each scalar degree of freedom could be a `(100, 100)` array, so my output is `(100, 100)`.

Where numpy's eager evaluation really hurts is dealing with automated gradient and Hessian computation. Empirically, calls to my generated Hessian function (using [autograd](https://github.com/HIPS/autograd)) end up being more than half my total computation time. I suspect this is due to the creation of spurious temporary objects. If a tool like Blaze was able to optimize this computational graph, that could be very helpful.

I have yet to find a solution for Hessians/gradients which (a) works for expressions generated at run-time, (b) allows for broadcasting input arguments of unknown (until run-time) shape, and (c) handles expression trees with thousands of ops well. Autograd+generated numpy is my current approach, which is not compatible with numba (see HIPS/autograd#47). I have been able to get numba to work if I first perform the differentiation in SymPy, but unfortunately this burns about two orders of magnitude more computation time than it saves (see [this comment](https://github.com/numba/numba/issues/1532#issuecomment-156858740)). Every other approach I've tried (e.g., Theano) has failed.

If Blaze solved this problem for me I would become a heavy user and big fan.

**Comment by datnamer** [comment]
@richardotis  Thanks for sharing your thoughts. 

CC: @mrocklin who was a sympy and blaze developer and might have ideas here.

**Comment by kwmsmith** [comment]
@datnamer @richardotis Thanks for your input and comments.  It's a fortunate synchronicity that you're thinking along these lines -- I've had multiple conversations in recent weeks regarding Blaze and things like Bayesian Inference, more general ML, etc.

For the foreseeable future, we are focused on making blaze more robust, more consistent (API-wise), better performance, more explorable, etc.  I have some near-term priorities that prevent me from spending too much time in the blaze ideas world, but once those are resolved I will look at this thread in more depth.

**Comment by datnamer** [comment]
@kwmsmith That sounds very exciting! I think blaze has alot of potential in this regard.

Here is another  front end DSL: http://people.csail.mit.edu/dhlin/jubayes/julia_bayes_inference.pdf 

Regarding generally ML etc in blaze ecosystem, some thoughts below from reading ideas being bounced around and thinking about abstractions in our own codebase (most of these are not my own, but make sense to me). I'll just dump them below:

For general but scalable inference Black Box Variation Bayes seems very promising. 

It would be awesome to have a modular interface with Model(distributions, nonparametrics etc)l->solvers ->data-> backend  etc that can be plugged in and out. 

Regarding blaze as part of a more general pipeline, some very instructive discussions in Julia land for abstractions here: https://github.com/JuliaStats/Roadmap.jl/issues  and here: https://github.com/Evizero/LearnBase.jl 

By Aside from different data sources, having modular abstractions in the middle will allow library writers to adjust algorithms and models without rewriting common parts of code, making research less arduous and encouraging it in the python ecosystem. 

This is even more general, but I'll dump my thoughts: A common theme that I'm seeing is   that having a unified type system for reuse is nice for some things, but can be too restrictive. On the other hand python duck typing is too loose, and I think the best balance  is to dispatch on behavior ie interfaces or protocols, when allowing for others to plug into your code. Python already has some shades of this in __array__ etc

https://github.com/python/typing/issues/11

Guido seems to be a big fan of this more regimented duck typing and I think it might go into follow up to the typing pep.

**Comment by njsmith** [comment]
> Examples of packages and constraints: Patsy is not easily extensible for things like operations and multiple variables on the left hand side of the formula and Pymc 3 is restricted in extending it for lme4 style mixed models. 

On the specific example of patsy, which sounds like it's probably a tangent from the main discussion here: the core is actually rather extensible, it's just that I haven't had time+motivating use cases to expose this via a proper public API, and no-one else has been interested enough to do it either. I think this is mostly orthogonal to blaze-related things though, because patsy is really a parser for an entirely different language and semantics than regular Python (though it embeds Python expressions as terminals in its parse tree). I'd love to see patsy become more extensible for problem where it'd be appropriate, like GAM or (generalized) mixed-effect models, but it's just not an appropriate tool if what you want to do is write arbitrary computational flows and then manipulate them in generic ways -- it has a more specialized area of application.

**Comment by datnamer** [comment]
Here is another good inspiration: http://autumnai.com/leaf/book/leaf.html

**Comment by datnamer** [comment]
So [Edward](https://github.com/blei-lab/edward) is a new (and looks to be promising) Probabilistic programming project by @dustinvtran and others  David Blei's Lab  at Columbia.

There is some discussion about designing a  model specification api:

https://github.com/blei-lab/edward/issues/27
https://github.com/blei-lab/edward/issues/17
https://github.com/blei-lab/edward/issues/20

This seems to dovetail with the above discussion and might be a great motivating case for a future blaze statistical/ML expression API.

**Comment by datnamer** [comment]
Lots of momentum and potential in the tensorflow world: http://dustintran.com/blog/a-quick-update-edward-and-some-motivations/ 

If this is still a project the blaze team want to take on, I think it would be best to boostrap off all the aforementioned progress. One hangup is not being able to write normal imperative code and being constrained to tensorflow abstractions.

I see two areas for participation:

1.  Perhaps there could be some way for Dynd/numba/blaze expresions to be exposed to tensorflow autodiff  and participate in tensorflow graph. 

2. Abstract/symbolic transformations  distributions and layers that can be shared  and multiple libraries, to unify the fragmentation. 

3. What about multiple backends ? (databases hdfs etc)

**Comment by llllllllll** [comment]
I think that anything more domain specific would need to be built as a project on top of blaze and not into blaze itself. Right now blaze is used for many things other than ML and I think that the team is too small to devote any meaningful amount of time to a project that is not activly solving our particular needs.

**Comment by datnamer** [comment]
Right, I thought so, but I wanted to dump my thoughts here since there was some interest expressed. 

I would be interested in helping build something on top of blaze, but I think there is still too much churn in the codebase and unanswered questions about defining new abstract and concrete types, integration with Dynd and the buffer protocol, api etc

**Comment by llllllllll** [comment]
I am not sure what you mean by, "too much churn in the codebase". The code has been pretty stable for a while now. I am also unsure how `dynd` integration or the buffer protocol are related to building new expressions into blaze.

**Comment by datnamer** [comment]
By churn, I meant (upcoming or perhaps not?) churn.

Here are some of the relevant issues I've been following: 

https://github.com/blaze/blaze/issues/776
https://github.com/blaze/blaze/issues/924
https://github.com/blaze/blaze/issues/923
https://github.com/blaze/blaze/issues/1402
https://groups.google.com/a/continuum.io/forum/#!topic/blaze-dev/PA1voi_6S6s

Regarding dynd, I was waiting to see how blaze expressions would play out with  Dynd's multiple dispatch system and ability to back arrays with different types of memory.
